lib/Sema/CSDiagnostics.cpp

```diff
@@ -703,6 +689,27 @@ bool GenericArgumentsMismatchFailure::diagnoseAsError() {
     }
   }
 
+  if (!diagnostic) {
+    // Handle all mismatches involving an `AssignExpr`
+    if (auto *assignExpr = dyn_cast<AssignExpr>(anchor)) {
+      diagnostic = getDiagnosticFor(CTP_AssignSource);
+      fromType = getType(assignExpr->getSrc());
+      toType = getType(assignExpr->getDest());
+    } else {
+      // If we couldn't find a specific diagnostic let's fallback to
+      // attempt to handle cases where we have an apply arg to param.
+      auto applyInfo = getFunctionArgApplyInfo(getLocator());
+      if (applyInfo) {
+        diagnostic = diag::cannot_convert_argument_value;
+        // When we have an InOutExpr, it can have inout to pointer
+        // conversions so we don't want the diagnose the inout type.
+        if (!isa<InOutExpr>(anchor))
+          fromType = applyInfo->getArgType();
```

Comment:
I'm not sure this condition is quite right, as it'll mean we'll diagnose things like:

```swift
func foo(_ x: inout [[Int]]) {}
func bar(_ x: inout [[String]]) {
  foo(&x)
}
```

as `cannot convert value of type '[String]' to expected argument type '[[Int]]'` (but the argument should really be reported as `[[String]]`).

I think for now we may want to keep the old logic for `case ConstraintLocator::OptionalPayload` to handle the inout-to-pointer case. Either that or pass the constraint restriction to `GenericArgumentsMismatch` when we create it in `simplifyRestrictedConstraintImpl`, and then check to see if we're dealing with an inout-to-pointer conversion when diagnosing.

In the long term, ideally the `ConstraintRestrictions` map would use a `ConstraintLocator` key which we could query, but that would require a major refactoring of CSApply.

Reply:
Humm, I didn't think of that case ... I'll adjust :) 
But I still think is worth to show the optional type for the param

Reply:
Fixed 👍